Core/clim-basic/extended-streams/recording.lisp

```diff
@@ -1131,7 +1131,11 @@ were added."
                 ;; (arguably) less bad than untrasnforming the coords back to
                 ;; how they were.
                 (let (,@(when (member 'coord-seq args)
-                          `((coord-seq (copy-seq coord-seq)))))
+                          `((coord-seq (copy-seq coord-seq))))
+                      ,@(when (member 'string args)
+                          `((string (subseq string (or start 0) (or end (length string))))
+                            (start  0)
+                            (end    nil))))
```

Comment:
I'd like rather medium-draw-text recording method be opencoded than picking up a single argument. Eventually, for this macro to scan all arguments for sequences and copy them without relying on the name.

That said this solution for strings is not complete, because output records may have strings added by means of the stream protocol, i.e add-string-output-to-text-record. When the string is styled, I think that it may be captured without "appending" to the existing string.

Reply:
> I'd like rather medium-draw-text recording method be opencoded than picking up a single argument.

Do you mean inlining? Or should the method be defined outside of the macro?

> Eventually, for this macro to scan all arguments for sequences and copy them without relying on the name.

I agree that looking at the name is pretty terrible, but I don't understand how the macro could decide which arguments will be sequences at runtime.

Reply:
> That said this solution for strings is not complete, because output records may have strings added by means of the stream protocol, i.e add-string-output-to-text-record. When the string is styled, I think that it may be captured without "appending" to the existing string.

I looked at `add-string-output-to-text-record` and it always copies the characters of the user-supplied string into an array owned by a `styled-string` so it should be fine. Are you aware of other places that keep references to user-supplied strings?

Reply:
you are right, in both arms of `if` we call replace. I'm not aware of other places.

Reply:
By open coding, I mean that the method should be defined outside of the macro. To inhibit the generation of the recording method there is a keyword argument ~medium-fn~ which needs to be specified as NIL.

I've meant that that the recording body could check the argument type at runtime and copy it on demand.

```
(defmethod medium-draw-text* ((medium recording) string)
  (let ((string (if (sequencep string) (copy-string …)))) …))
```
but that's indeed a terrible way to do things. So I think that the method should be written without using the macro. Draw-rectangle inhibits this method (notice, that it is defined as the ~around~ method to allow actual drawing).

Reply:
the call to the macro generate-medium-recording-body may be done inside the method of course, after copying the sequence. Since we are at it, we probably should do the same for coord-seq instead of relying on the argument name.